### PR TITLE
Add compatibility with FOSRest 2.0

### LINKED ISSUE #393: Problem with api
When i try to get the products using this url
`http://127.0.0.1:8000/api/ecommerce/products?page=1`
I got the following error

```
Catchable Fatal Error: Argument 2 passed to FOS\RestBundle\Util\ContextHelper::addGroups() 
must be of the type array, string given, called in /Users/carvajal/development/sukkyservices/bottleappostle/ba/vendor/friendsofsymfony/rest-bundle/FOS/RestBundle/EventListener/ViewResponseListener.php 
on line 106 and defined
```

### LINKED ISSUE #437: update composer error with Sonata\CustomerBundle
Hello All
I have this error when i update the composer

[Doctrine\Common\Annotations\AnnotationException]                          
    [Creation Error] The annotation @QueryParam declared on method Sonata\Cus  
  tomerBundle\Controller\Api\AddressController::getAddressesAction() does not  
   have a property named "array". Available properties: requirements, map, al  
  lowBlank, name, key, default, description, strict, nullable, incompatibles   
                                                                               
                 Thanks

Joseph

### LINKED ISSUE #454: Error in Api\BasketController::getBasketsAction()
#### Sonata packages

```bash
sonata-project/admin-bundle              3.20.1             The missing Symfony Admin Generator
sonata-project/block-bundle              3.3.2              Symfony SonataBlockBundle
sonata-project/cache                     1.0.7              Cache library
sonata-project/cache-bundle              2.3.1              This bundle provides caching services
sonata-project/classification-bundle     3.3.1              Symfony SonataClassificationBundle
sonata-project/comment-bundle            3.0.0              Integrate the FOSCommentBundle in the Sonata Project
sonata-project/core-bundle               3.4.0              Symfony SonataCoreBundle
sonata-project/dashboard-bundle          dev-master 1d62f37 Provides a Dashboard management through container and block services
sonata-project/datagrid-bundle           2.2.1              Symfony SonataDatagridBundle
sonata-project/doctrine-extensions       1.0.2              Doctrine2 behavioral extensions
sonata-project/doctrine-orm-admin-bundle 3.1.5              Symfony Sonata / Integrate Doctrine ORM into the SonataAdminBundle
sonata-project/easy-extends-bundle       2.2.0              Symfony SonataEasyExtendsBundle
sonata-project/ecommerce                 2.x-dev 57edf4a    E-Commerce solution provided by Sonata
sonata-project/exporter                  1.7.1              Lightweight Exporter library
sonata-project/formatter-bundle          3.2.2              Symfony SonataFormatterBundle
sonata-project/google-authenticator      2.0.0              Library to integrate Google Authenticator into a PHP project
sonata-project/intl-bundle               2.3.1              Symfony SonataIntlBundle
sonata-project/media-bundle              3.5.1              Symfony SonataMediaBundle
sonata-project/news-bundle               3.x-dev 7496755    Symfony SonataNewsBundle
sonata-project/notification-bundle       3.1.0              Symfony SonataNotificationBundle
sonata-project/page-bundle               3.x-dev 0716a11    This bundle provides a Site and Page management through container and block services
sonata-project/seo-bundle                2.2.0              Symfony SonataSeoBundle
sonata-project/sonata-composer           dev-master 6ebe1f3
sonata-project/timeline-bundle           3.1.1              Integrates SpyTimelineBundle into Sonata
sonata-project/translation-bundle        dev-master a6c11c6 SonataTranslationBundle
sonata-project/user-bundle               3.2.4              Symfony SonataUserBundle
```

#### Symfony packages

```bash
symfony/assetic-bundle     v2.8.1            Integrates Assetic into Symfony2
symfony/monolog-bundle     v2.12.1           Symfony MonologBundle
symfony/phpunit-bridge     2.7.x-dev 29fa7b8 Symfony PHPUnit Bridge
symfony/polyfill-apcu      v1.4.0            Symfony polyfill backporting apcu_* functions to lower PHP versions
symfony/polyfill-intl-icu  v1.4.0            Symfony polyfill for intl's ICU-related data and classes
symfony/polyfill-mbstring  v1.4.0            Symfony polyfill for the Mbstring extension
symfony/polyfill-php54     v1.4.0            Symfony polyfill backporting some PHP 5.4+ features to lower PHP versions
symfony/polyfill-php55     v1.4.0            Symfony polyfill backporting some PHP 5.5+ features to lower PHP versions
symfony/polyfill-php56     v1.4.0            Symfony polyfill backporting some PHP 5.6+ features to lower PHP versions
symfony/polyfill-php70     v1.4.0            Symfony polyfill backporting some PHP 7.0+ features to lower PHP versions
symfony/polyfill-util      v1.4.0            Symfony utilities for portability of PHP codes
symfony/security-acl       v3.0.0            Symfony Security Component - ACL (Access Control List)
symfony/swiftmailer-bundle v2.6.2            Symfony SwiftmailerBundle
symfony/symfony            v2.8.22           The Symfony PHP framework
```

#### PHP version

```bash
PHP 7.0.0 (cli) (built: Dec  8 2015 16:39:19) ( NTS )
Copyright (c) 1997-2015 The PHP Group
Zend Engine v3.0.0, Copyright (c) 1998-2015 Zend Technologies
```

## Subject

nelmio/api-doc-bundle 2.x-dev 17b6c3a Generates documentation for your REST API from annotations

Fatal error: Uncaught Doctrine\Common\Annotations\AnnotationException: [Creation Error] The annotation @QueryParam declared on method Sonata\BasketBundle\Controller\Api\BasketController::getBasketsAction() does not have a property named "array". Available properti     es: requirements, map, allowBlank, name, key, default, description, strict, nullable, incompatibles in /Users/Stephane/Sites/Falciform/vendor/doctrine/annotations/lib/Doctrine/Common/Annotations/AnnotationException.php:69        

## Steps to reproduce

composer update

### LINKED ISSUE #520: Move api groups from string to array
Example: https://github.com/sonata-project/SonataMediaBundle/pull/1155

Note: `orderBy` is fixed in #519

### PR DESCRIPTION
**Pull request by Xymanek**
I am targeting this branch, because this is a BC layer for 2.x branch.

Closes #454 
Closes #437
Closes #520
Closes #393 

## Changelog

```markdown
### Fixed
- Compatibility with FOSRest 2.0
```

## Subject

See https://github.com/sonata-project/ecommerce/pull/519#issuecomment-365075541

### CODE REVIEW

**Comment by greg0ire** [comment]
I think @jordisala1991 listed more issues than that, didn't he?

**Comment by Xymanek** [comment]
I checked and they seem to be unrelated

File: src/BasketBundle/Controller/Api/BasketController.php [hunk 1]
```diff
@@ -96,6 +95,21 @@ public function __construct(BasketManagerInterface $basketManager, BasketElement
      */
     public function getBasketsAction(ParamFetcherInterface $paramFetcher)
     {
+        $orderByQueryParam = new QueryParam();
+        $orderByQueryParam->name = 'orderBy';
+        $orderByQueryParam->requirements = 'ASC|DESC';
+        $orderByQueryParam->nullable = true;
+        $orderByQueryParam->strict = true;
+        $orderByQueryParam->description = 'Query baskets basket by clause (key is field, value is direction)';
```
Comment (greg0ire) [inline src/BasketBundle/Controller/Api/BasketController.php]: Please take this as an occasion to change "Query baskets basket by" to "Basket query order by", this looks silly.

File: src/CustomerBundle/Controller/Api/AddressController.php [hunk 2]
```diff
@@ -69,6 +68,21 @@ public function __construct(AddressManagerInterface $addressManager, FormFactory
      */
     public function getAddressesAction(ParamFetcherInterface $paramFetcher)
     {
+        $orderByQueryParam = new QueryParam();
+        $orderByQueryParam->name = 'orderBy';
+        $orderByQueryParam->requirements = 'ASC|DESC';
+        $orderByQueryParam->nullable = true;
+        $orderByQueryParam->strict = true;
+        $orderByQueryParam->description = 'Query orders addresses by clause (key is field, value is direction)';
```
Comment (greg0ire) [inline src/CustomerBundle/Controller/Api/AddressController.php]: This one is even weirder lol. Should be 
> Address query order by clause (key is field, value is direction)

File: src/CustomerBundle/Controller/Api/CustomerController.php [hunk 3]
```diff
@@ -86,6 +85,21 @@ public function __construct(CustomerManagerInterface $customerManager, OrderMana
      */
     public function getCustomersAction(ParamFetcherInterface $paramFetcher)
     {
+        $orderByQueryParam = new QueryParam();
+        $orderByQueryParam->name = 'orderBy';
+        $orderByQueryParam->requirements = 'ASC|DESC';
+        $orderByQueryParam->nullable = true;
+        $orderByQueryParam->strict = true;
+        $orderByQueryParam->description = 'Query customers order by clause (key is field, value is direction)';
```
Comment (greg0ire) [inline src/CustomerBundle/Controller/Api/CustomerController.php]: Customer query order by clause

File: src/InvoiceBundle/Controller/Api/InvoiceController.php [hunk 4]
```diff
@@ -59,6 +58,21 @@ public function __construct(InvoiceManagerInterface $invoiceManager)
      */
     public function getInvoicesAction(ParamFetcherInterface $paramFetcher)
     {
+        $orderByQueryParam = new QueryParam();
+        $orderByQueryParam->name = 'orderBy';
+        $orderByQueryParam->requirements = 'ASC|DESC';
+        $orderByQueryParam->nullable = true;
+        $orderByQueryParam->strict = true;
+        $orderByQueryParam->description = 'Query invoices invoice by clause (key is field, value is direction)';
```
Comment (greg0ire) [inline src/InvoiceBundle/Controller/Api/InvoiceController.php]: You know what to do

File: src/BasketBundle/Controller/Api/BasketController.php (hunk 1 above)
Comment (Xymanek) [reply src/BasketBundle/Controller/Api/BasketController.php]: Maybe "Basket query order**ed** by"?

Comment (Xymanek) [reply src/BasketBundle/Controller/Api/BasketController.php]: Or maybe like "Sort results by (key is field, value is direction)"

Comment (greg0ire) [reply src/BasketBundle/Controller/Api/BasketController.php]: > Sort results by (key is field, value is direction)

Feels less SQL-ish, might be better indeed 🤔

Comment (greg0ire) [reply src/BasketBundle/Controller/Api/BasketController.php]: Maybe just "sort specification (key is field, value is direction)" ?

Comment (greg0ire) [reply src/BasketBundle/Controller/Api/BasketController.php]: "sort specification for the current resultset (key is field, value is direction)"?

File: src/BasketBundle/Controller/Api/BasketController.php [hunk 5]
```diff
@@ -86,7 +86,6 @@ public function __construct(BasketManagerInterface $basketManager, BasketElement
      *
      * @QueryParam(name="page", requirements="\d+", default="1", description="Page for baskets list pagination (1-indexed)")
      * @QueryParam(name="count", requirements="\d+", default="10", description="Number of baskets by page")
-     * @QueryParam(name="orderBy", array=true, requirements="ASC|DESC", nullable=true, strict=true, description="Query baskets basket by clause (key is field, value is direction")
      *
      * @View(serializerGroups="sonata_api_read", serializerEnableMaxDepthChecks=true)
```
Comment (jordisala1991) [inline src/BasketBundle/Controller/Api/BasketController.php]: Can you change `"sonata_api_read"` by `{"sonata_api_read"}`?

So we fix all the issues at once? It should be simple. It should be also changed on the ApiDoc above

Comment (Xymanek) [reply src/BasketBundle/Controller/Api/BasketController.php]: I was going to ask about this - should I do it in this PR or in a separate one considering that `orderBy` is fixed differently in master?

Comment (jordisala1991) [reply src/BasketBundle/Controller/Api/BasketController.php]: separate commit would be good enough

Comment (Xymanek) [reply src/BasketBundle/Controller/Api/BasketController.php]: 👍

File: src/BasketBundle/Controller/Api/BasketController.php (hunk 1 above)
Comment (Xymanek) [reply src/BasketBundle/Controller/Api/BasketController.php]: Maybe "sort specification for the ~current~ resultset (key is field, value is direction)"

> Sort specification for the resultset (key is field, value is direction)

**Comment by jordisala1991** [comment]
I have edited your PR message, added the label and reviewed the code.

Last thing missing is fix the build

**Comment by Xymanek** [comment]
Damn.... I knew something will go wrong by not following the interface... 

How is #393 related?

**Comment by jordisala1991** [comment]
Probably because the groups are a string before your PR, and after an array. And the guy is talking about an exception on `addGroups` receiving an string instead of an array.

It has a high chances to be related

File: src/BasketBundle/Controller/Api/BasketController.php [hunk 6]
```diff
@@ -81,21 +81,35 @@ public function __construct(BasketManagerInterface $basketManager, BasketElement
...
-     *  output={"class"="Sonata\DatagridBundle\Pager\PagerInterface", "groups"="sonata_api_read"}
+     *  output={"class"="Sonata\DatagridBundle\Pager\PagerInterface", "groups"={"sonata_api_read"}}
      * )
      *
      * @QueryParam(name="page", requirements="\d+", default="1", description="Page for baskets list pagination (1-indexed)")
      * @QueryParam(name="count", requirements="\d+", default="10", description="Number of baskets by page")
-     * @QueryParam(name="orderBy", array=true, requirements="ASC|DESC", nullable=true, strict=true, description="Query baskets basket by clause (key is field, value is direction")
      *
-     * @View(serializerGroups="sonata_api_read", serializerEnableMaxDepthChecks=true)
+     * @View(serializerGroups="{sonata_api_read"}, serializerEnableMaxDepthChecks=true)
```
Comment (jordisala1991) [inline src/BasketBundle/Controller/Api/BasketController.php]: Careful, it should be: {"sonata_api_read"} instead of "{sonata_api_read"}

Comment (jordisala1991) [reply src/BasketBundle/Controller/Api/BasketController.php]: it happens everywhere I believe

Comment (Xymanek) [reply src/BasketBundle/Controller/Api/BasketController.php]: Ah, dammit. Yeah, I just mass-replaced everything...

**Comment by Xymanek** [comment]
Small problem while fixing tests:

```php
    public function testGetBasketsAction()
    {
        $basketManager = $this->createMock(BasketManagerInterface::class);
        $basketManager->expects($this->once())->method('getPager')->will($this->returnValue([]));

        $paramFetcher = $this->createMock(ParamFetcherInterface::class);
        $paramFetcher->expects($this->exactly(3))->method('get');
        $paramFetcher->expects($this->once())->method('all')->will($this->returnValue([]));
        $paramFetcher->expects($this->once())->method('addParam')->with($this->callback(function ($param) {
            return $param instanceof QueryParam && $param->name = 'orderBy';
        }));

        $this->assertEquals([], $this->createBasketController($basketManager)->getBasketsAction($paramFetcher));
    }
```

```
Testing Sonata\BasketBundle\Tests\Controller\Api\BasketControllerTest

Trying to configure method "addParam" which cannot be configured because it does not exist, has not been specified, is final, or is static
```

Any suggestions on what to do?

**Comment by jordisala1991** [comment]
Do you know any class implementing that interface, and having that method? Maybe you can create mock of that instead.

**Comment by Xymanek** [comment]
Can someone restart travis build? It timed out....

**Comment by jordisala1991** [comment]
Thank you @Xymanek

**Pull request merged**